src/input.rs

```diff
@@ -70,6 +70,48 @@ where
             changed_at: ChangedAt::Revision(Revision::ZERO),
         })
     }
+
+    fn set_common(&self, key: &Q::Key, value: Q::Value, changed_at: impl FnOnce() -> ChangedAt) {
+        let map = self.map.upgradable_read();
```

Comment:
Why do we need two-phase read/write locking here? I think that setting has to be exclusive, so we can just grab the write lock?

Reply:
Revised. In the process, I also fixed some latent bugs. See latest commit.